core/src/main/scala/chisel3/Aggregate.scala

```diff
@@ -604,6 +665,66 @@ object VecInit extends SourceInfoDoc {
   def do_fill[T <: Data](n: Int)(gen: => T)(implicit sourceInfo: SourceInfo, compileOptions: CompileOptions): Vec[T] =
     apply(Seq.fill(n)(gen))
 
+  /** Creates a new 2D [[Vec]] of length `n by m` composed of the result of the given
+   * function applied to an element of data type T.
+   * 
+   * @param n number of inner vectors (rows) in the outer vector
+   * @param m number of elements in each inner vector (column)
+   * @param gen function that takes in an element T and returns an output 
+   * element of the same type
+   */
+  def fill[T <: Data](n: Int, m: Int)(gen: => T): Vec[Vec[T]] = macro VecTransform.fill2D
+
+  /** @group SourceInfoTransformMacro */
+  def do_fill[T <: Data](n: Int, m: Int)(gen: => T)(implicit sourceInfo: SourceInfo, compileOptions: CompileOptions): Vec[Vec[T]] = {
+    val myVec = Wire(Vec(n, Vec(m, cloneSupertype(Seq.fill(m)(gen), "Vec"))))
+    myVec.foreach {
+      _ := VecInit.fill(m)(gen)
+    }
+    myVec
+  }
+
+  /** Creates a new 3D [[Vec]] of length `n by m by p` composed of the result of the given
+   * function applied to an element of data type T.
+   * 
+   * @param n number of 2D vectors inside outer vector
+   * @param m number of 1D vectors in each 2D vector
+   * @param p number of elements in each 1D vector
+   * @param gen function that takes in an element T and returns an output 
+   * element of the same type
+   */
+  def fill[T <: Data](n: Int, m: Int, p: Int)(gen: => T): Vec[Vec[Vec[T]]] = macro VecTransform.fill3D
+
+  /** @group SourceInfoTransformMacro */
+  def do_fill[T <: Data](n: Int, m: Int, p: Int)(gen: => T)(implicit sourceInfo: SourceInfo, compileOptions: CompileOptions): Vec[Vec[Vec[T]]] = {
+    val myVec = Wire(Vec(n, Vec(m, Vec(p, cloneSupertype(Seq.fill(p)(gen), "Vec")))))
+    myVec.foreach { 
+      _ := VecInit.fill(m, p)(gen)
+    }
+    myVec
+  }
+
+  /** Creates a new 4D [[Vec]] of length `n by m by p by q` composed of the result of the given
```

Comment:
is there any particular reason you went to 4D, not 3D or 5D? Is this paralleling something else in the codebase and/or scala?

Reply:
Scala seqs can go to 4D, so I think its to follow that pattern (https://www.scala-lang.org/api/2.13.3/scala/collection/SeqFactory.html)

Reply:
Ah, I See that scala Seq.tabulate and so on have up to 5D versions. But there they don't have different names, they just have different numbers of parameters to distinguish... any reason to not follow that pattern

https://www.scala-lang.org/api/2.13.6/scala/collection/Seq$.html#tabulate[A](n1:Int,n2:Int,n3:Int,n4:Int,n5:Int)(f:(Int,Int,Int,Int,Int)=%3EA):CC[CC[CC[CC[CC[A]]]]@scala.annotation.unchecked.uncheckedVariance]

Reply:
I guess I might not encourage people to have 5D versions in hardware, and might stop at 3D for chisel... but willing to be convinced otherwise

Reply:
Ah oh I guess I forgot that it went to 5D, @jackkoenig suggested to go to 4D and I assumed thats kinda where it came from... I could add a 5D if you want it?

Reply:
i guess i'd be more inclined to go lower, not higher, but I'd defer to @jackkoenig on what he thinks is best. I'm not sure how realistic 5-d generated hardware would be / how happy tools like VCS would be about trying to do something with that

Reply:
I was thinking 4D because it's what `Seq` did in Scala 2.12. I agree that 5D is a bit much and maybe 4D is too much. Maybe let's go to 3D for the time being?